File: services/src/Moe.La.Core/Entities/Litigation/Case/Case.cs
using Moe.La.Core.Enums;
using System;
using System.Collections.Generic;

namespace Moe.La.Core.Entities
{
    public class Case : BaseEntity<int>
    {
        /// <summary>
        /// مصدر القضية
        /// </summary>
        public CaseSources CaseSource { get; set; }

        /// <summary>
        /// رقم القضية في المصدر
        /// </summary>
        public string CaseNumberInSource { get; set; }

        /// <summary>
        /// سنة القضية في المصدر
        /// </summary>
        public int CaseYearInSource { get; set; }

        /// <summary>
        /// درجة الترافع
        /// </summary>
        public LitigationTypes LitigationType { get; set; }

        /// <summary>
        /// رقم القضية المرجعية
        /// </summary>
        /// <remarks>هو رقم يتم توليده من خلال النظام</remarks>
        public string ReferenceCaseNo { get; set; }

        /// <summary>
        /// الرقم الموحد
        /// </summary>
        /// <remarks>هو رقم المسئول عن ربط القضايا المرتبطة ببعضها البعض، يتم توليده من خلال النظام</remarks>
        public string UnifiedNumber { get; set; }

        /// <summary>
        /// تاريخ بداية القضية
        /// </summary>
        public DateTime StartDate { get; set; }

        /// <summary>
        /// رقم معرف المحكمة
        /// </summary>
        public int? CourtId { get; set; }

        /// <summary>
        /// المحكمة
        /// </summary>
        public Court Court { get; set; }

        /// <summary>
        /// رقم الدائرة
        /// </summary>
        public string CircleNumber { get; set; }

        /// <summary>
        /// عنوان الدعوى
        /// </summary>
        public string Subject { get; set; }

        /// <summary>
        /// موضوع الدعوى
        /// </summary>
        public string CaseDescription { get; set; }

        /// <summary>
        /// صفة الوزارة القانونية
        /// </summary>
        public MinistryLegalStatuses LegalStatus { get; set; }

        /// <summary>
        /// القضية السابقة المرتبطة
        /// </summary>
        public int? RelatedCaseId { get; set; }

        /// <summary> 
        /// The related case.
        /// </summary>
        /// <remarks>Used to chain the related cases.</remarks>
        public Case RelatedCase { get; set; }

        /// <summary>
        /// The status of the case
        /// </summary>
        public CaseStatuses Status { get; set; }

        /// <summary>
        /// The legal affair department related to this case.
        /// </summary>
        public int BranchId { get; set; }

        /// <summary>
        /// The legal affair department object.
        /// </summary>
        public Branch Branch { get; set; }

        /// <summary>
        /// After the case is closed, it will be archived.
        /// </summary>
        public bool IsArchived { get; set; } = false;

        /// <summary>
        /// رقم ملف القضية
        /// </summary>
        public string FileNo { get; set; }

        /// <summary>
        /// اسم القاضي
        /// </summary>
        public string JudgeName { get; set; }

        /// <summary>
        /// تاريخ إغلاق القضية
        /// </summary>
        public DateTime? CloseDate { get; set; }

        /// <summary>
        /// تاريخ النطق بالحكم
        /// </summary>
        public DateTime? PronouncingJudgmentDate { get; set; }

        /// <summary>
        /// موعد استلام الحكم
        /// </summary>
        public DateTime? ReceivingJudgmentDate { get; set; }

        /// <summary>
        /// ملاحظات
        /// </summary>
        public string Notes { get; set; }

        /// <summary>
        /// The user's id who created it.
        /// </summary>
        public Guid CreatedBy { get; set; }

        /// <summary>
        /// Navigation property to the user who created it.
        /// </summary>
        public AppUser CreatedByUser { get; set; }

        /// <summary>
        /// The user's id who updated it.
        /// </summary>
        public Guid? UpdatedBy { get; set; }

        /// <summary>
        /// Navigation property to the user who updated it.
        /// </summary>
        public AppUser UpdatedByUser { get; set; }

        /// <summary>
        /// The update datetime.
        /// </summary>
        public DateTime? UpdatedOn { get; set; }

        /// <summary>
        /// Case rule based on the final hearing of type "نطق بالحكم"
        /// </summary>
        public CaseRule CaseRule { get; set; }

        /// <summary>
        /// تصنيف القضية
        /// </summary>
        public int SecondSubCategoryId { get; set; }

        /// <summary>
        /// Second Sub Category "التصنيف الفرعى 2"
        /// </summary>

        public SecondSubCategory SecondSubCategory { get; set; }

        /// <summary>
        /// سبب إغلاق القضية
        /// </summary>
        public CaseClosinReasons? CaseClosingReason { get; set; }

        /// <summary>
        /// يدوى
        /// </summary>
        public bool IsManual { get; set; } = false;

        /// <summary>
        /// مرفقات اسانيد الدعوى
        /// </summary>
        public ICollection<CaseGrounds> CaseGrounds { get; set; } = new List<CaseGrounds>();

        /// <summary>
        /// جلسات القضية
        /// </summary>
        public ICollection<Hearing> Hearings { get; set; } = new List<Hearing>();

        /// <summary>
        /// اطراف القضية
        /// </summary>
        public ICollection<CaseParty> Parties { get; set; } = new List<CaseParty>();

        /// <summary>
        /// الباحثين المكلفين على القضية
        /// </summary>
        public ICollection<CaseResearcher> Researchers { get; set; } = new List<CaseResearcher>();

        /// <summary>
        /// حركات القضية
        /// </summary>
        public ICollection<CaseTransaction> CaseTransactions { get; set; } = new List<CaseTransaction>();

        /// <summary>
        /// مرفقات القضية
        /// </summary>
        public ICollection<CaseAttachment> Attachments { get; set; } = new List<CaseAttachment>();

        /// <summary>
        /// طلبات الخصوم
        /// </summary>
        public ICollection<CaseRuleProsecutorRequest> ProsecutorRequests { get; set; } = new List<CaseRuleProsecutorRequest>();

        /// <summary>
        /// السجل التاريخي للقضية
        /// </summary>
        public ICollection<CaseHistory> CaseHistory { get; set; } = new List<CaseHistory>();

        /// <summary>
        /// المعاملات المرتبطة
        /// </summary>
        public ICollection<CaseMoamala> CaseMoamalat { get; set; } = new List<CaseMoamala>();


        #region Properties likely will be deleted in the future
        ///// <summary>
        ///// رقم القيد في راسل
        ///// </summary>
        //public string RaselRef { get; set; }

        ///// <summary>
        ///// الرقم الموحد في راسل
        ///// </summary>
        //public string RaselUnifiedNo { get; set; }

        ///// <summary>
        ///// رقم المعاملة في ناجز
        ///// </summary>
        //public string NajizRef { get; set; }

        ///// <summary>
        ///// تاريخ المعاملة في ناجز
        ///// </summary>
        //public DateTime? NajizRefDate { get; set; }

        ///// <summary>
        ///// رقم الطلب في ناجز
        ///// </summary>
        //public string NajizId { get; set; }

        ///// <summary>
        ///// رقم المعاملة في معين
        ///// </summary>
        //public string MoeenRef { get; set; }

        ///// <summary>
        ///// تاريخ المعاملة في معين
        ///// </summary>
        //public DateTime? MoeenRefDate { get; set; }

        ///// <summary>
        ///// وصف الطلب
        ///// </summary>
        //public string OrderDescription { get; set; }

        ///// <summary>
        ///// The related workflow instance id.
        ///// </summary>
        //public Guid? WorkflowInstanceId { get; set; }

        //public WorkflowInstance WorkflowInstance { get; set; }

        //public DateTime? RecordDate { get; set; }
        #endregion
    }
}

File: website/Moe.La.Web/els-app/src/app/core/models/case.ts
import { KeyValuePairs } from './key-value-pairs';
import { BaseModel } from './base-model';
import { Attachment } from './attachment';
import { CaseRuleDetails } from './case-rule';
import { CasePartyDetails } from './case-party';
import { HearingDetails } from './hearing';

export class CaseListItem extends BaseModel<number> {
  id: number = 0;
  receivedDate: string = ''; // تاريخ الاستلام
  receivedDateHigri: string = ''; // تاريخ الاستلام
  createdOnHigri: string = ''; // تاريخ الاستلام

  receiptStatus: string = ''; // حالة الاستلام
  raselRef: string = '';
  raselUnifiedNo: string = '';
  caseNumberInSource: string = '';//رقم القضية
  najizRef: string = ''; // رقم المعاملة ٝي ناجز

  najizId: string = ''; // رقم الطلب ٝي ناجز
  moeenRef: string = ''; // رقم المعاملة ٝي معين

  caseSource!: KeyValuePairs; //مصدر القضية
  litigationType: string = ''; // درجة التراٝع
  mainNo: string = ''; // الرقم الرئيسي
  startDate: string = ''; // تاريخ بداية القضية
  startDateHigri: string = ''; // تاريخ بداية القضية
  court: string = ''; //المحكمة
  circleNumber: string = ''; //الدائرة
  subject: string = ''; //عنوان الدعوى
  type: string = ''; //صٝة الوزارة القانونية
  caseDescription: string = ''; //موضوع الدعوى
  attachments: string[] = []; // مرٝقات أسانيد الدعوى[قائمة]
  orderDescription: string = ''; //وصٝ الطلب

  isDuplicated?: boolean; // مكررة أم لا
  relatedCaseId?: number; // القضية المرتبطة
  relatedCase!: CaseListItem;
  branch?: string; // الإدارة المحال إليها القضية
  status!: KeyValuePairs; // حالة القضية
  hearingsCount?: number; // عدد الجلسات ٝي القضية

  pronouncingJudgmentDate: string = ''; // تاريخ نطق الحكم
  receivingJudgmentDate: string = '';// موعد استلام الحكم
  caseGrounds: string[] = [];
  pronouncingJudgmentDateHigri?: string = '';
  receivingJudgmentDateHigri?: string = '';
  objectionJudgmentLimitDateHijri?: string = '';
  isAllowObjectionMemo?: boolean = false;
  secondSubCategoryId: number;
  notes: string = ''; // ملاحظات

  caseSourceNumber: string = '';

}

export class CaseDetails extends BaseModel<number> {
  isManual?: boolean;
  raselRef: string = '';
  raselUnifiedNo: string = '';
  najizRef: string = '';
  najizId: string = '';
  moeenRef: string = '';
  caseNumberInSource: string = '';
  parentId?: number;
  caseSource: KeyValuePairs = { id: 0, name: '' };
  litigationType: KeyValuePairs = { id: 0, name: '' };
  referenceCaseNo: string = '';
  mainNo: string = '';
  startDateHigri: string = '';
  startDate!: Date;
  court: KeyValuePairs = { id: 0, name: '' };
  circleNumber: string = '';
  subject: string = '';
  legalStatus: KeyValuePairs = { id: 0, name: '' };
  caseDescription: string = '';
  orderDescription: string = '';
  status: KeyValuePairs = { id: 0, name: '' };
  branchId?: number = 0;
  branchName: string = '';
  recordDate: string = '';
  fileNo: string = '';
  judgeName: string = '';
  closeDate: string = '';
  relatedCaseId?: number;
  relatedCase!: CaseListItem;
  pronouncingJudgmentDate?: Date;
  receivingJudgmentDate?: Date;
  caseRule?: CaseRuleDetails;
  caseRuleId?: number;
  judgementText?: string;
  isFinalJudgment?: boolean;
  judgementResult?: string;
  judgmentBrief?: string;
  attachmentsCount!: number;
  hearings: any[] = null!;
  caseMoamalat?: any[] = [];
  researchers: KeyValuePairs<string>[] = null!;
  parties: any[] = [];
  caseParties?: CasePartyDetails[] = [];
  plaintiffs: any[] = [];
  respondents: any[] = [];
  judgementImportDate?: Date;
  objectionJudgmentLimitDateHijri?: any;
  pronouncingJudgmentDateHijri?: any;
  receivingJudgmentDateHijri?: any;

  caseTransactions?: CaseTransaction[] = [];
  secondSubCategory?: KeyValuePairs = { id: 0, name: '' };
  subCategory?: any;
  caseGrounds: CaseGrounds[] = [];
  prosecutorRequests: any[] = [];
  attachments: any[] = [];
  notes: string = '';
  finishedPronouncedHearing:boolean;
  remainingObjetcion:number=0;
}

export class CaseTransaction extends BaseModel<number> {

  createdByUser: string = '';
  // createdByRole: string = '';
  text: string = '';

}

export class CaseGrounds {
  text: string = '';
}

export class SaveCase {
  id: number = 0;
  raselRef: string = '';
  raselUnifiedNo: string = '';
  caseSourceNumber: string = '';
  najizRef: string = '';
  najizId: string = '';
  moeenRef: string = '';
  caseSource: string = '';
  litigationType: string = '';
  referenceCaseNo: string = '';
  mainNo: string = '';
  startDate!: Date;
  courtId: number = 0
  circleNumber: string = '';
  subject: string = '';
  legalStatus: string = '';
  caseDescription: string = '';
  orderDescription: string = '';
  status: string = '';
  branchId?: number = 0;
  recordDate?: string = '';
  fileNo: string = '';
  judgeName: string = '';
  //closeDate?: string = '';
  //pronouncingJudgmentDate?: Date;
  receivingJudgmentDate?: Date;

  caseGrounds: CaseGrounds[] = [];
  attachments?: Attachment[];
  notes: string = '';
}

export class ObjectionCase {

  Id: number = 0;
  caseSource: number = 0;
  caseSourceNumber: string = '';
  courtId: number = 0;
  circleNumber: string = '';
  relatedCaseId: number;
  LitigationType: number

}

export class MainCaseDetails {
  id: number = 0;
  raselRef: string = '';
  raselUnifiedNo: string = '';
  caseNumberInSource: string = '';
  najizRef: string = '';
  najizId: string = '';
  moeenRef: string = '';
  caseSource: KeyValuePairs = { id: 0, name: '' };
  litigationType: KeyValuePairs = { id: 0, name: '' };
  referenceCaseNo: string = '';
  mainNo: string = '';
  startDate!: Date;
  courtId: number = 0
  circleNumber: string = '';
  subject: string = '';
  legalStatus: KeyValuePairs = { id: 0, name: '' };
  caseDescription: string = '';
  orderDescription: string = '';
  status: KeyValuePairs = { id: 0, name: '' };
  branchId?: number = 0;
  recordDate?: string = '';
  fileNo: string = '';
  judgeName: string = '';
  subCategory?: SecondSubCategory;
  receivingJudgmentDate?: Date;
  court: KeyValuePairs = { id: 0, name: '' };
  caseParties?: CasePartyDetails[] = [];
  caseGrounds?: any[] = [];
  caseMoamalat?: any[] = [];
  attachments?: Attachment[] = [];
  hearings?: HearingDetails[] = [];
  notes: string = '';
  relatedCaseId?: number;
  startDateHigri: string = '';
  relatedCase!: CaseListItem;
  moamalaId?: number;
}

export class SecondSubCategory {
  id: number;
  name: string;
  FirstSubCategoryId: number;
  firstSubCategory: FirstSubCategory;
  mainCategory: MainCategory;
}

export class FirstSubCategory {
  id: number;
  name: string;
  MainCategoryId: number;
}

export class MainCategory {
  id: number;
  name: string;
}
